internal/remote/remote_catalog.go

```diff
@@ -102,9 +124,103 @@ func (c *RemoteCatalog) Delete(
 ) error {
 	skrContext, err := c.skrContextFactory.Get(kyma)
 	if err != nil {
-		return fmt.Errorf("failed to get SkrContext for deleting RemoteCatalog: %w", err)
+		return fmt.Errorf("failed to get SKR context: %w", err)
 	}
 
 	moduleTemplates := c.moduleTemplateSyncAPIFactoryFn(c.kcpClient, skrContext.Client, &c.settings)
 	return moduleTemplates.DeleteAllManaged(ctx)
 }
+
+// GetModuleReleaseMetasToSync returns a list of ModuleReleaseMetas that should be synced to the SKR.
+// A ModuleReleaseMeta that is Beta or Internal is synced only if the Kyma is also Beta or Internal.
+func (c *RemoteCatalog) GetModuleReleaseMetasToSync(
+	ctx context.Context,
+	kyma *v1beta2.Kyma,
+) ([]v1beta2.ModuleReleaseMeta, error) {
+	moduleReleaseMetaList := &v1beta2.ModuleReleaseMetaList{}
+	if err := c.kcpClient.List(ctx, moduleReleaseMetaList); err != nil {
+		return nil, fmt.Errorf("failed to list ModuleReleaseMetas: %w", err)
+	}
+
+	moduleReleaseMetas := []v1beta2.ModuleReleaseMeta{}
+	for _, moduleReleaseMeta := range moduleReleaseMetaList.Items {
+		if moduleReleaseMeta.IsBeta() && !kyma.IsBeta() {
+			continue
+		}
+		if moduleReleaseMeta.IsInternal() && !kyma.IsInternal() {
+			continue
+		}
+		moduleReleaseMetas = append(moduleReleaseMetas, moduleReleaseMeta)
+	}
+
+	return moduleReleaseMetas, nil
+}
+
+// GetModuleTemplatesToSync returns a list of ModuleTemplates that should be synced to the SKR.
+// A ModuleTemplate is synced if it is not mandatory and does not have sync disabled. In addition,
+// it must be referenced by a ModuleReleaseMeta that is synced.
+func (c *RemoteCatalog) GetModuleTemplatesToSync(
+	ctx context.Context,
+	moduleReleaseMetas []v1beta2.ModuleReleaseMeta,
+) ([]v1beta2.ModuleTemplate, error) {
+	moduleTemplateList := &v1beta2.ModuleTemplateList{}
+	if err := c.kcpClient.List(ctx, moduleTemplateList); err != nil {
+		return nil, fmt.Errorf("failed to list ModuleTemplates: %w", err)
+	}
+
+	return c.FilterModuleTemplatesToSync(moduleTemplateList.Items, moduleReleaseMetas), nil
+}
+
+func (c *RemoteCatalog) FilterModuleTemplatesToSync(
+	moduleTemplates []v1beta2.ModuleTemplate,
+	moduleReleaseMetas []v1beta2.ModuleReleaseMeta,
+) []v1beta2.ModuleTemplate {
+	moduleTemplatesToSync := map[string]bool{}
+	for _, moduleReleaseMeta := range moduleReleaseMetas {
+		for _, channel := range moduleReleaseMeta.Spec.Channels {
+			moduleTemplatesToSync[fmt.Sprintf("%s-%s", moduleReleaseMeta.Spec.ModuleName, channel.Version)] = true
+		}
+	}
+
+	filteredModuleTemplates := []v1beta2.ModuleTemplate{}
+	for _, moduleTemplate := range moduleTemplates {
+		if moduleTemplate.IsMandatory() {
+			continue
+		}
+
+		if moduleTemplate.HasSyncDisabled() {
+			continue
+		}
+
+		if _, found := moduleTemplatesToSync[moduleTemplate.Name]; found {
```

Comment:
I am thinking if this is safe. Here we're matching the module based on it's `metadata.Name`. Wouldn't it be more robust to use module's `spec.ModuleName + "-" + spec.Version`? I know these two should be in-sync, but what if they're not?
What do you think?

Reply:
good point, better save than sorry. Adapted here: https://github.com/kyma-project/lifecycle-manager/pull/2097/commits/dfbc32e57d623b0b9d3e64ca12796efce1ca85c1